File: frontend/src/lib/requests/approvals.ts
import client from "../client";
import { Approval } from "../models/Approval";

export async function checkUserApproval(eventId: string, userId: string): Promise<Approval> {
  const endpoint = `/approvals/${eventId}/${userId}`;

  return (await client.get<Approval>(endpoint)).data
}

export async function requestApproval(userId: string, eventId: string) {
  const endpoint = `/approvals/${eventId}/${userId}`;

  await client.post(endpoint)
}

export async function loadApprovalRequests(eventId: string): Promise<Approval[]> {
  const endpoint = `/approvals/${eventId}/`;

  return (await client.get(endpoint)).data
}

export async function approveUser(eventId: string, userId: string) {
  const endpoint = `/approvals/${eventId}/${userId}`;

  await client.patch(endpoint, { approved: true })
}

export async function disapproveUser(eventId: string, userId: string) {
  const endpoint = `/approvals/${eventId}/${userId}`;

  await client.patch(endpoint, { approved: false })
}

export async function resetApproval(eventId: string, userId: string) {
  const endpoint = `/approvals/${eventId}/${userId}`;

  await client.patch(endpoint, { approved: null })
} 

File: backend/src/routes/approvals.py
from fastapi import APIRouter, Depends
from src.models.approval import Approval
from src.services.approvals import ApprovalService
from src.containers import Container
from dependency_injector.wiring import inject, Provide


router = APIRouter(
    prefix="/approvals",
    dependencies=[
        Depends(Provide[Container.approvals_service]),
    ],
)


@router.get("/{event_id}/{user_id}")
@inject
async def getApprovals(
    event_id,
    user_id,
    approval_service: ApprovalService = Depends(Provide[Container.approvals_service]),
):
    return approval_service.get(user_id, event_id)

@router.patch("/{event_id}/{user_id}")
@inject
async def updateApproval(
    event_id,
    user_id,
    approval: Approval,
    approval_service: ApprovalService = Depends(Provide[Container.approvals_service]),
):
    approval_service.update(event_id, user_id, approval.approved)

@router.post("/{event_id}/{user_id}")
@inject
async def createApprovalRequest(
    event_id,
    user_id,
    approval_service: ApprovalService = Depends(Provide[Container.approvals_service]),
):
    return approval_service.create(event_id, user_id)


@router.get("/{event_id}/")
@inject
async def getApprovalRequestsForEvent(
    event_id,
    approval_service: ApprovalService = Depends(Provide[Container.approvals_service]),
):
    return approval_service.get_by_event_id(event_id)


